# Part 1. Equation of a Slime

How many late days are you using for this assignment? 1

In [466]:
# Imports section
import pandas as pd
import sklearn
import sklearn.model_selection as skmd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
import numpy as np
from sklearn.preprocessing import PolynomialFeatures

## 1. Loading the dataset

In [467]:
# Using pandas load the dataset
data = pd.read_csv('science_data_large.csv')
# Output the first 15 rows of the data
print(data.head(15))
# Display a summary of the table information (number of datapoints, etc.)

print("\nSummary of the table:\n")
data.info()

rows, columns = data.shape
print(f"\nNumber of data points (rows): {rows}")
print(f"Number of features (columns): {columns - 1}")
print("Number of target columns: 1")

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04

Summary of the table:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2 

## 2. Splitting the dataset

In [468]:
# Take the pandas dataset and split it into our features (X) and label (y)
targetCol = 'Size nm^3'
X = data.drop(columns=[targetCol])
y = data[targetCol]

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

## 3. Perform a Linear Regression

In [469]:
# Use sklearn to train a model on the training set
model = LinearRegression()
model.fit(X_train, y_train)

# Create a sample datapoint and predict the output of that sample with the trained model
samplePoint = pd.DataFrame({"Temperature °C": [400], "Mols KCL": [600]})
prediction = model.predict(samplePoint)
print("Size nm^3 prediction for \n", samplePoint, "\n is: ", prediction, "\n")

# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means
score = model.score(X_test, y_test)
print("Score: ", score)

# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
coefficients = model.coef_
intercept = model.intercept_
print(f"\nCoefficients are: {[f'{coef:.5f}' for coef in coefficients]}")
print(f"Intercept is: {intercept:.5f}")

Size nm^3 prediction for 
    Temperature °C  Mols KCL
0             400       600 
 is:  [556684.13] 

Score:  0.8552472077276095

Coefficients are: ['866.14641', '1032.69507']
Intercept is: -409391.47958


Write the linear equation of a slime: (example equation: $E = mc^2$)

$h(x) = -409391.47958 + 866.14641 \cdot Temperature + 1032.69507 \cdot Mols KCL$

Report on score and explain meaning:

The score for the linear regression model is 0.85525. The score is the coefficient of determination of the prediction. The best possible score is 1.0 which would indict that the model perfectly predicts the outcome. Therefore the score indicates how well the features explain the variability of the target variable (Size nm^3) and the closer to 1, the better fit a model is to the data. Since the score is 0.85525, this model is a decently good fit to the data but not a perfect fit.

## 4. Use Cross Validation

In [470]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42
kFold = skmd.KFold(n_splits=5, shuffle=True, random_state=42)
scores = skmd.cross_val_score(model, X, y, cv=kFold)
# Report on their finding and their significance
print(f"Cross-validated scores: {[f'{score:.5f}' for score in scores]}")
print("Mean: ", np.mean(scores))
print("Std: ", np.std(scores))

Cross-validated scores: ['0.86152', '0.82742', '0.87195', '0.88166', '0.85609']
Mean:  0.8597294202684646
Std:  0.01838773713930639


Write findings here:

The 5 fold cross-validated scores were 0.86152, 0.82742, 0.87195, 0.88166, 0.85609. The mean is 0.85973 and the standard deviation is 0.01839. The model is a decently good fit for the data since the all the scores are around 0.85 however it is not a perfect fit since the scores are not 1.0. The small deviation indicts that the model is relatively stable across the data because the value only fluctuate slightly when using different data. Since the R2s are consistent, likely no overfitting because it works on multiple different data splits.

## 5. Using Polynomial Regression

In [471]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)
poly = PolynomialFeatures(degree=2)
X_poly = poly.fit_transform(X)

feature_names = poly.get_feature_names_out()

# ????
linear_model = LinearRegression()


kFold = skmd.KFold(n_splits=5, shuffle=True, random_state=42)
scores = skmd.cross_val_score(linear_model, X_poly, y, cv=kFold)
# Report on their finding and their significance
print("Cross-validated scores:", scores)
print("Mean: ", np.mean(scores))
print("Std: ", np.std(scores))

linear_model.fit(X_poly, y)

coefficients = linear_model.coef_
intercept = linear_model.intercept_
print("\nCoefficient labels are: ", feature_names)
print(f"\nCoefficients are: {[f'{coef:.5f}' for coef in coefficients]}")
print(f"Intercept is: {intercept:.5f}")

Cross-validated scores: [1. 1. 1. 1. 1.]
Mean:  1.0
Std:  0.0

Coefficient labels are:  ['1' 'Temperature °C' 'Mols KCL' 'Temperature °C^2'
 'Temperature °C Mols KCL' 'Mols KCL^2']

Coefficients are: ['0.00000', '12.00000', '-0.00000', '-0.00000', '2.00000', '0.02857']
Intercept is: 0.00002


Write the polynomial equation of a slime: (example equation: $E = mc^2$)

$Size = 0.0 + 12 \cdot Temperature - 0.0 \cdot Mols KCL - 0.0 \cdot Temperature C^2 + \\
 2 \cdot Temperature \cdot Mols KCL + 0.02857 \cdot Mols KCL^2$

Report on the score and interpret:

Since all the scores are 1.0, the model perfectly fits the data. A score of 1.0 indicates that that model can correctly classify the data perfectly. Since all the data splits have a score of 1.0, this model is perfectly fit for this data. It is not just overfit for the training data.

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [472]:
# Load the dataset. Then train and evaluate the classification models.
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier

ckdData = pd.read_csv('ckd_feature_subset.csv')

X = ckdData.drop(columns=['Target_ckd'])
y = ckdData['Target_ckd']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

logReg_model = LogisticRegression()
svc_model = SVC()
kNN_model = KNeighborsClassifier()
neural_model = MLPClassifier()

logReg_model.fit(X_train, y_train)
svc_model.fit(X_train, y_train)
kNN_model.fit(X_train, y_train)
neural_model.fit(X_train, y_train)

logReg_scores = skmd.cross_val_score(logReg_model, X, y, cv=5)
svc_scores = skmd.cross_val_score(svc_model, X, y, cv=5)
kNN_scores = skmd.cross_val_score(kNN_model, X, y, cv=5)
neural_scores = skmd.cross_val_score(neural_model, X, y, cv=5)

models = ["Logistic Regression", "Support Vector Machines", "k-Nearest Neighbors", "Neural Network"]
avg_accuracies = [np.mean(logReg_scores), np.mean(svc_scores), np.mean(kNN_scores), np.mean(neural_scores)]
std_devs = [np.std(logReg_scores), np.std(svc_scores), np.std(kNN_scores), np.std(neural_scores)]

results_df = pd.DataFrame({
    'Model': models,
    'Average Accuracy': avg_accuracies,
    'Standard Deviation': std_devs
})

print(results_df)

/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iteration

                     Model  Average Accuracy  Standard Deviation
0      Logistic Regression          0.856129            0.054236
1  Support Vector Machines          0.928172            0.059993
2      k-Nearest Neighbors          0.928172            0.031893
3           Neural Network          0.888817            0.054190


/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


## Results and Conclusion for Classification Experiments

The models with higher average accuracies are better fit to the data, therefore the svc model and kNN model have the best fit since they have the highest average accuracy of 0.928172. The next best fit is the neural networks model with an accuracy of 0.869247 and last is logistic regression with and accuracy of 0.856129. While the svc model and kNN model had the same average accuracy/fit to the data, the kNN model had a slightly lower standard deviation which means it was more consistent across different datasets which makes it a slightly better model choice than the svc model.

According to these results, the kNN model performed the best because it had the highest accuracy and lowest standard deviation. That means of all the models it has the best fit for the data while also having the least variability between different datasets.

In [473]:
# Experiments with Neural Network.

In [474]:
smallLayer = MLPClassifier(hidden_layer_sizes=(20))
biggerLayer = MLPClassifier(hidden_layer_sizes=(100))
twoLayers = MLPClassifier(hidden_layer_sizes=(50, 50))
relu = MLPClassifier(activation='relu')
logistic = MLPClassifier(activation='logistic')
tanh = MLPClassifier(activation='tanh')
smallMaxIter = MLPClassifier(hidden_layer_sizes=(50, 50), max_iter=50)
largeMaxIter = MLPClassifier(hidden_layer_sizes=(50, 50), max_iter=300)

models = [
    ("Small layer", smallLayer),
    ("Bigger layer", biggerLayer),
    ("Two layers", twoLayers),
    ("Relu", relu),
    ("Logistic", logistic),
    ("Tanh", tanh),
    ("Small max iter", smallMaxIter),
    ("Big max iter", largeMaxIter)
]

config_names = []
avg_accuracies = []
std_devs = []

for conf_name, model in models:
    scores = skmd.cross_val_score(model, X, y, cv=5)
    avg_accuracy = np.mean(scores)
    std_dev = np.std(scores)
    
    config_names.append(conf_name)
    avg_accuracies.append(avg_accuracy)
    std_devs.append(std_dev)

# neuralnet1_scores = skmd.cross_val_score(neuralnet1, X, y, cv=5)
# neuralnet2_scores = skmd.cross_val_score(neuralnet2, X, y, cv=5)
# neuralnet3_scores = skmd.cross_val_score(neuralnet3, X, y, cv=5)
# neuralnet4_scores = skmd.cross_val_score(neuralnet4, X, y, cv=5)
# neuralnet5_scores = skmd.cross_val_score(neuralnet5, X, y, cv=5)
# neuralnet6_scores = skmd.cross_val_score(neuralnet6, X, y, cv=5)
# neuralnet7_scores = skmd.cross_val_score(neuralnet7, X, y, cv=5)

# models = ["neuralnet1", "neuralnet2", "neuralnet3", "neuralnet4", "neuralnet5", "neuralnet6", "neuralnet7"]
# avg_accuracies = [np.mean(neuralnet1_scores), np.mean(neuralnet2_scores), np.mean(neuralnet3_scores), np.mean(neuralnet4_scores), np.mean(neuralnet5_scores), np.mean(neuralnet6_scores), np.mean(neuralnet7_scores)]
# std_devs = [np.std(neuralnet1_scores), np.std(neuralnet2_scores), np.std(neuralnet3_scores), np.std(neuralnet4_scores), np.std(neuralnet5_scores), np.std(neuralnet6_scores), np.std(neuralnet7_scores)]

results_nn = pd.DataFrame({
    'Model': config_names,
    'Average Accuracy': avg_accuracies,
    'Standard Deviation': std_devs
})

print(results_nn)

/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iteration

            Model  Average Accuracy  Standard Deviation
0     Small layer          0.849677            0.054005
1    Bigger layer          0.888817            0.057903
2      Two layers          0.947957            0.033076
3            Relu          0.882151            0.049926
4        Logistic          0.830108            0.064387
5            Tanh          0.888817            0.057903
6  Small max iter          0.849677            0.054005
7    Big max iter          0.954194            0.016173


/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (300) reached and the optimization hasn't converged yet.
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (300) reached and the optimization hasn't converged yet.
  warnings.warn(


## Results and Conclusion for Neural Network Experiments

The more neurons in a layer, it seems the better fit the model is to the data since the the accuracy increases from 0.849677 to 0.881935 as the number of neurons increase. Also, the more layers there are, the better the fit since with two layers instead of one the accuracy increases from 0.85ish to over 0.9. Furthermore the accuracy of the model also changes when the activation function changes. Of three activation functions: relu, logistic, and tanh, tanh has the highest accuracy (0.888602), then relu (0.862581), and finally logistic (0.817204). Therefore, using the tanh function provides the best fit for the data compared to the other activation functions. Lastly, having a bigger max iteration leads to a better fit to the data. When the neural network model has a max iteration of 50 then the model only has an average accuracy of 0.849677 which is much lower than the model that has a max iteration of 300, which has an accuracy of 0.960645 and a lower standard deviation. This means the model with a bigger max iteration has an almost perfect fit to the data.

The two parameters that had the biggest impact was the number of layers and the number of max iterations. With only one layer the accuracy was only in the mid-high 0.8s whereas with an additional layer the accuracy increases to closer to 0.95. That is much closer to being a perfect fit for the data. Therefore, you can see that if a model has more layers, it seems to better fit the data. Changing the number of max iterations from 50 to 300 increases the accuracy by 0.1 which is the biggest change in the table. Therefore, this parameter has a big impact on the classification process and can lead to a much better fit for the data. Also another relevant parameter is the activation function. In this case, the tanh function had the best fit for the data especially compared to the logistic function which performed much worse.